modules/ipp/scanner.go

```diff
@@ -225,80 +211,149 @@ func bufferFromBody(res *http.Response, scanner *Scanner) *bytes.Buffer {
+		valueTag := interim[len(interim)-3]
 		var vals [][]byte
-		buf := bytes.NewBuffer(interims[1])
+		buf := bytes.NewBuffer(interims[i+1])
 		// This reading occurs in a loop because some attributes can have type "1 setOf <type>"
 		// where same attribute has a set of values, rather than one
 		for tag, nameLength := valueTag, int16(0); tag == valueTag && nameLength == 0; {
 			var length int16
 			if err := binary.Read(buf, binary.BigEndian, &length); err != nil {
 				//Couldn't read length of content
-				return vals, err
+				break
 			}
 			val := make([]byte, length)
 			if err := binary.Read(buf, binary.BigEndian, &val); err != nil {
 				//Couldn't read content
-				vals = append(vals, val)
-				return vals, err
+				//vals = append(vals, val)
+				break
 			}
 			vals = append(vals, val)
 			if err := binary.Read(buf, binary.BigEndian, &tag); err != nil {
 				//Couldn't read next valueTag
-				return vals, err
+				break
 			}
 			// FIXME: Only try to read next namelength if previous valueTag wasn't end-of-attributes-tag
 			if err := binary.Read(buf, binary.BigEndian, &nameLength); err != nil {
 				//Couldn't read next nameLength
-				return vals, err
+				break
 			}
 		}
-		return vals, nil
+		valSlice = append(valSlice, vals)
 	}
-	//The attribute was not present
-	return nil, errors.New("Attribute \"" + attrString + "\" not present.")
+	return valSlice, nil
 }
 
-func (scan *scan) tryReadAttributes(body string) {
-	bodyBytes := []byte(body)
-	// Write reported CUPS version to results object
-	if scan.results.AttributeCUPSVersion == "" {
-		if cupsVersions, err := readAttributeFromBody(CupsVersion, &bodyBytes); err != nil {
-			log.WithFields(log.Fields{
-				"error":     err,
-				"attribute": CupsVersion,
-			}).Debug("Failed to read attribute.")
-		} else if len(cupsVersions) > 0 {
-			scan.results.AttributeCUPSVersion = string(cupsVersions[0])
+type Value struct {
+	Bytes []byte `json:"raw,omitempty"`
+}
+
+type Attribute struct {
+	Name string    `json:"name,omitempty"`
+	Values []Value `json:"values,omitempty"`
+	ValueTag byte  `json:"tag,omitempty"`
+}
+
+type AttributeGroup struct {
+	Attrs []*Attribute
+	Tag byte
+}
+
+func printGroup(g *AttributeGroup) {
+	fmt.Printf("0x%x\n", g.Tag)
+	for _, attrP := range g.Attrs {
+		fmt.Printf("\t0x%x: %q\n", attrP.ValueTag, attrP.Name)
+		for _, v := range attrP.Values {
+			fmt.Printf("\t\t%v\n", v)
 		}
 	}
-	// Write reported IPP versions to results object
-	if len(scan.results.AttributeIPPVersions) == 0 {
-		if ippVersions, err := readAttributeFromBody(VersionsSupported, &bodyBytes); err != nil {
-			log.WithFields(log.Fields{
-				"error":     err,
-				"attribute": VersionsSupported,
-			}).Debug("Failed to read attribute.")
-		} else {
-			for _, v := range ippVersions {
-				scan.results.AttributeIPPVersions = append(scan.results.AttributeIPPVersions, string(v))
+}
+
+// TODO: Log every error that could come out of this
```

Comment:
Some errors off the top of my head --

 * Data too short (an immediate check that `len(body) > sizeof(start) + sizeof(tag) + sizeof(len)`, or if you get an io.EOF when you're not expecting it)
 * Invalid data (e.g. if you get a length that is longer than the remaining data, or if you can't access the length of the remaining data, longer than `len(body)`)
 * Read error (the only one that may need special handling is `io.EOF`, otherwise I don't see why they couldn't just be returned as-is)